### PR TITLE
Fix UB in atomics with automatic storage

### LINKED ISSUE #2585: [BUG]: Invalid results with automatic storage when using `std::atomic`
### Is this a duplicate?

- [x] I confirmed there appear to be no [duplicate issues](https://github.com/NVIDIA/cccl/issues) for this bug and that I agree to the [Code of Conduct](CODE_OF_CONDUCT.md)

### Type of Bug

Runtime Error

### Component

libcu++

### Describe the bug

Iniitially #478 intended to fix this, but the code has diverged beyond the point that this can be cleanly merged.

Using `cuda::std::atomic` with thread local memory in device code has the potential to invoke a reduction across the entire warp and return invalid results when doing single increments.

### How to Reproduce

```c++
template <typename T>
__device__ T fetch_add(T in) {
  cuda::atomic<T> x = in;
  x.fetch_add(T(1), cuda::memory_order_relaxed);
  return x.load(cuda::memory_order_relaxed);
}
```

### Expected behavior

The fix is to detect if the atomic is operating on local memory, for 12.3 and older, this behavior is disabled as there are performance issues with `__islocal`.

### Reproduction link

_No response_

### Operating System

_No response_

### nvidia-smi output

_No response_

### NVCC version

_No response_

### PR DESCRIPTION
**Pull request by wmaxey**
## Description

Closes #2585 

See #478 for more details.

This fixes UB when using automatic storage with atomics (which has little or no use-case)

Example of users encountering wrong values in automatic storage: https://stackoverflow.com/questions/75640847/what-exactly-does-cudaatomic-does

## Checklist

- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+_CCCL_DEVICE bool __cuda_store_weak_if_local(volatile _Type* __ptr, _Type __val)
+{
+#if defined(_LIBCUDACXX_ATOMIC_UNSAFE_AUTOMATIC_STORAGE)
+  if (!__cuda_is_local((const void*) __ptr))
+  {
+    return false;
+  }
+  memcpy((void*) __ptr, (void const*) &__val, sizeof(_Type));
+  return true;
+#else
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h:115]: Missing the comments

File: libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    return false;
+  }
+  memcpy((void*) __ret, (void const*) __ptr, sizeof(_Type));
+  // Required to workaround a compiler bug, see nvbug/4064730
+  __nanosleep(0);
+  return true;
+#else
+  return false;
+#endif
+}
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h:103]: Question: This is only needing the size of `_Type`

Could we pass in a void* and either template this on size_t or pass it as an argument?

Applies below

File: libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+template <class _Type>
+_CCCL_DEVICE bool
+__cuda_compare_exchange_weak_if_local(volatile _Type* __ptr, _Type* __expected, const _Type* __desired, bool* __success)
+{
+  if (!__cuda_is_local((const void*) __ptr))
```
Comment (gonzalobg) [inline libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h:124]: We are missing the `#if defined(_LIBCUDACXX_ATOMIC_UNSAFE_AUTOMATIC_STORAGE)`. 
I'd expect that we'd want it here for the same reason we want it in the other ops.

File: libcudacxx/include/cuda/std/detail/libcxx/include/__config [hunk 4]
```diff
@@ -792,6 +792,11 @@ __sanitizer_annotate_contiguous_container(const void*, const void*, const void*,
 #    define _LIBCUDACXX_HAS_MSVC_ATOMIC_IMPL
 #  endif
 
+// Enable bypassing automatic storage checks in atomics when using CTK 12.2 and below and if NDEBUG is defined.
+#  if defined(_CCCL_CUDACC_BELOW_12_2) && !defined(NDEBUG)
+#    define _LIBCUDACXX_ATOMIC_UNSAFE_AUTOMATIC_STORAGE
+#  endif // _CCCL_CUDACC_BELOW_12_2
```
Comment (gonzalobg) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__config:798]: @gevtushenko fyi (I remember you had feedback about this)

File: libcudacxx/codegen/generators/compare_and_swap.h [hunk 5]
```diff
@@ -154,6 +154,8 @@ static inline _CCCL_DEVICE bool __atomic_compare_exchange_cuda(_Type* __ptr, _Ty
   __proxy_t* __ptr_proxy = reinterpret_cast<__proxy_t*>(__ptr);
   __proxy_t* __exp_proxy = reinterpret_cast<__proxy_t*>(__exp);
   __proxy_t* __des_proxy  = reinterpret_cast<__proxy_t*>(&__des);
+  bool __res = false;
+  if (__cuda_compare_exchange_weak_if_local(__ptr_proxy, __exp_proxy, __des_proxy, &__res)) {{return __res;}}
```
Comment (ericniebler) [inline libcudacxx/codegen/generators/compare_and_swap.h:158]: probably a n00b question, but why the double braces `{{}}`?

Comment (wmaxey) [reply libcudacxx/codegen/generators/compare_and_swap.h:158]: Double brackets are an escape for `fmt`, they likely just leaked in. Thanks!

**Comment by wmaxey** [comment]
I've finally taken another pass at this. I moved the logic for bypassing the checks into `__cuda_is_local` so it's all managed from one location. Most every function checks it as a prologue.

File: libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h [hunk 6]
```diff
@@ -0,0 +1,209 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h:7]:
```suggestion
// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
```

File: libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+#if _CCCL_HAS_CUDA_COMPILER
+
+_CCCL_DEVICE inline bool __cuda_is_local(const volatile void* __ptr)
+{
+#  if defined(_LIBCUDACXX_ATOMIC_UNSAFE_AUTOMATIC_STORAGE)
+  return false;
+// Only NVCC+NVRTC define __isLocal, so drop to PTX
+#  elif defined(_CCCL_CUDACC_BELOW_12_3) || defined(_CCCL_CUDA_COMPILER_NVHPC)
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h:42]: We have new macros for that
```suggestion
#  elif _CCCL_CUDACC_BELOW(12,3) || _CCCL_CUDA_COMPILER(NVHPC)
```

File: libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+  int __tmp = 0;
+  asm("{\n\t"
+      "  .reg .pred p;\n\t"
+      "  isspacep.local p, %1;\n\t"
+      "  @p mov.s32 %0, 1;\n\t"
+      "}\n\t"
+      : "=r"(__tmp)
+      : "l"(const_cast<const void*>(__ptr)));
+  return __tmp == 1;
+#  else
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h:52]:
```suggestion
#  else // ^^^ _CCCL_CUDACC_BELOW(12,3) || _CCCL_CUDA_COMPILER(NVHPC) ^^^ / vvv other compiler vvv
```

File: libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  return __cuda_fetch_weak_if_local(__ptr, __val, __ret, __cuda_fetch_local_bop_max<_Type>);
+}
+
+template <class _Type>
+_CCCL_DEVICE bool __cuda_fetch_min_weak_if_local(volatile _Type* __ptr, _Type __val, _Type* __ret)
+{
+  return __cuda_fetch_weak_if_local(__ptr, __val, __ret, __cuda_fetch_local_bop_min<_Type>);
+}
+
+#endif
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h:205]:
```suggestion
#endif // _CCCL_HAS_CUDA_COMPILER
```

File: libcudacxx/include/cuda/std/detail/libcxx/include/__config [hunk 10]
```diff
@@ -737,6 +737,11 @@ typedef unsigned int char32_t;
 #    define _LIBCUDACXX_HAS_MSVC_ATOMIC_IMPL
 #  endif
 
+// Enable bypassing automatic storage checks in atomics when using CTK 12.2 and below and if NDEBUG is defined.
+#  if defined(_CCCL_CUDACC_BELOW_12_2) && !defined(NDEBUG)
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__config:741]:
```suggestion
#  if _CCCL_CUDACC_BELOW(12,2) && !defined(NDEBUG)
```

File: libcudacxx/include/cuda/std/detail/libcxx/include/__config [hunk 11]
```diff
@@ -737,6 +737,11 @@ typedef unsigned int char32_t;
...
+#    define _LIBCUDACXX_ATOMIC_UNSAFE_AUTOMATIC_STORAGE
+#  endif // _CCCL_CUDACC_BELOW_12_2
```
Comment (miscco) [inline libcudacxx/include/cuda/std/detail/libcxx/include/__config:743]:
```suggestion
#  endif // _CCCL_CUDACC_BELOW(12,2)
```

**Review by miscco — CHANGES_REQUESTED** [review]

File: libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+      "  .reg .pred p;\n\t"
+      "  isspacep.local p, %1;\n\t"
+      "  @p mov.s32 %0, 1;\n\t"
+      "}\n\t"
+      : "=r"(__tmp)
+      : "l"(const_cast<const void*>(__ptr)));
+  return __tmp == 1;
+#  else
+  return __isLocal(const_cast<const void*>(__ptr));
+#  endif
```
Comment (miscco) [inline libcudacxx/include/cuda/std/__atomic/functions/cuda_local.h:54]:
```suggestion
#  endif // _CCCL_CUDACC_AT_LEAST(12,3) && !_CCCL_CUDA_COMPILER(NVHPC)
```

**Review by miscco — APPROVED** [review]

**Review by miscco — APPROVED** [review]

**Pull request merged**